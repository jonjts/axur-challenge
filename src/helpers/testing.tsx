import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';

import ThemeProvider from '../providers/Theme';
import InspectionProvider from '../providers/Inspection'

export const putProviders = (component: JSX.Element) => (
    <ThemeProvider>
        <InspectionProvider>
            {component}
        </InspectionProvider>
    </ThemeProvider>
);

export const matchSnapshotWithProviders = (component: JSX.Element) => {
    const tree = renderer.create(putProviders(component)).toJSON();
    expect(tree).toMatchSnapshot();
    return tree;
};

export const renderWithProviders = (component: JSX.Element) => {
    return render(putProviders(component));
};
