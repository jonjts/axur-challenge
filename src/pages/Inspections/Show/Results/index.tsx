import React, { memo, useMemo } from 'react';
import { useTheme } from 'styled-components';

import { InspectionResult, INSPECTION_RESULT_STATUS } from '../../../../entities';

import Button from '../../../../components/Button';

import {
  StatusContainer,
  Indicator,
  StatusName,
  ResultList,
  ListItem
} from './styles';

type Props = {
  result: InspectionResult
};

const Show: React.FC<Props> = ({ result }) => {
  const { colors } = useTheme();

  const colorStatus = useMemo(() => result.status === INSPECTION_RESULT_STATUS.active
    ? colors.secondary
    : colors.success
    , [colors, result.status]);

  const handleOpenResult = (url: string) => {
    window.open(url, '_blank');
  };

  return (
    <>
      <StatusContainer>
        <div className='row'>
          <div className='column'>
            <label>
              Status
            </label>
            <StatusName className='row'>
              <Indicator
                color={colorStatus}
              />
              <label>
                {result.status}
              </label>
            </StatusName>
          </div>
        </div>
      </StatusContainer>
      <ResultList>
        <h2>Resultados</h2>
        {
          !result.urls.length ? (
            <p>Nenhum resultado encontrado</p>
          ) : (
            result.urls.map(url => (
              <ListItem key={url}>
                <Button
                  isLink
                  onClick={() => handleOpenResult(url)}
                >
                  {url}
                </Button>
              </ListItem>
            ))
          )
        }
      </ResultList>
    </>
  );
};

export default memo(Show);