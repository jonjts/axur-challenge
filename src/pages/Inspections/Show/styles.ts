import styled from 'styled-components';

export const CardHeader = styled.div`
    display: flex;
    flex: 1;
    flex-direction: row;
    align-items: center;
    border-bottom: 1px solid ${({ theme: { colors } }) => colors.default};

    button {
        margin-left: ${({ theme: { spacings: { sm } } }) => sm.pixel};
    }

    @media (min-width: 0) { 
        flex-direction: column;
        max-height: 200px;

        button {
            margin-bottom: ${({ theme: { spacings: { sm } } }) => sm.pixel};
        }
    }

    @media (min-width: 992px) { 
        flex-direction: row;
        max-height: 70px;

        button {
            margin-bottom: 0;
        }
    }
`;

export const CardBody = styled.div`
    flex: 1;
    min-height: 200px;
    margin-top: ${({ theme: { spacings: { xl } } }) => xl.pixel};
`;

export const Content = styled.div`
    width: 100%;

    @media (min-width: 992px) { 
        grid-template-columns: 50% 50% !important;
    }
`;

export const SpinnerContainer = styled.div`
    flex: 1;
    justify-content: center;
    align-items: center;
    flex-direction: row;
`;
