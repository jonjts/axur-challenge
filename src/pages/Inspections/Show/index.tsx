import React, { useState, useEffect, useCallback } from 'react';
import { useHistory, useParams } from 'react-router-dom';

import { InspectionResult } from '../../../entities';
import { useInspection } from '../../../providers/Inspection';

import Spinner from '../../../components/Spinner';
import Card from '../../../components/Card';
import Button from '../../../components/Button';

import Results from './Results';

import {
  CardHeader,
  CardBody,
  SpinnerContainer,
  Content
} from './styles';

const Show: React.FC = () => {
  const { goBack } = useHistory();
  const { find, loading } = useInspection();
  const { id } = useParams() as { id?: string };

  const [inspectionId, setInspectionId] = useState<string>(id || '');
  const [inspectionResult, setInspectionResult] = useState({} as InspectionResult);

  const findInspection = useCallback(async () => {
    setInspectionResult({} as InspectionResult);
    const result = await find(inspectionId.trim());
    setInspectionResult(result);
  }, [find, inspectionId]);

  useEffect(() => {
    if (id) findInspection();
  }, [id]);

  return (
    <>
      <h1>Ver Inspeção</h1>
      <Card
        showBackButton
        backText='Voltar'
        onBackClicked={goBack}
        disableButtons={loading}
      >
        <CardHeader>
          {
            !id ? (
              <>
                <input
                  value={inspectionId}
                  onChange={({ target: { value } }) => setInspectionId(value)}
                  placeholder='Informe o id da inspeção'
                />
                <Button onClick={findInspection}>Pesquisar</Button>
              </>
            ) : (
              <h2>
                ID: <i>{id}</i>
              </h2>
            )
          }
        </CardHeader>
        <CardBody className='row'>
          {
            loading ? (
              <SpinnerContainer className='row'>
                <Spinner />
              </SpinnerContainer>
            ) : (
              <Content className='grid'>
                {inspectionResult?.id && <Results result={inspectionResult} />}
              </Content>
            )
          }
        </CardBody>
      </Card>
    </>
  );
};

export default Show;