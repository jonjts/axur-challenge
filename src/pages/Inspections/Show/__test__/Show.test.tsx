import { matchSnapshotWithProviders, renderWithProviders } from '../../../../helpers/testing';

import Show from '../index';

describe('Show Inspection', () => {
  test('match with snapshot', () => {
    matchSnapshotWithProviders(<Show />);
  });

  test('renders correctly', () => {
    renderWithProviders(<Show />);
  });
});
