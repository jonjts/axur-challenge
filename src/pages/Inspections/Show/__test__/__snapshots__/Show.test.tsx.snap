// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Show Inspection match with snapshot 1`] = `
Array [
  <h1>
    Ver Inspeção
  </h1>,
  <div
    className="sc-eCssSg fcCMAw"
  >
    <div
      className="sc-dQppl iUgCqc"
    >
      <input
        onChange={[Function]}
        placeholder="Informe o id da inspeção"
        value=""
      />
      <button
        className="sc-dlfnbm ciWBfn"
        data-testid="btn"
        onClick={[Function]}
      >
        Pesquisar
      </button>
    </div>
    <div
      className="sc-bqyKva ILfCF row"
    >
      <div
        className="sc-kstrdz jbgVNi grid"
      />
    </div>
    <div
      className="sc-jSgupP jxuqsq"
    >
      <button
        className="sc-hKgILt dcQvag"
        data-testid="btn-link"
      >
        Voltar
      </button>
    </div>
  </div>,
]
`;
